cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs

```diff
@@ -375,13 +441,89 @@ writeLedgerState mOutFile qState@(SerialisedLedgerState serLedgerState) =
     Just (OutputFile fpath) ->
       handleIOExceptT (ShelleyQueryCmdWriteFileError . FileIOError fpath)
         $ LBS.writeFile fpath $ unSerialised serLedgerState
- where
-   decodeLedgerState
-     :: SerialisedLedgerState era
-     -> Either LBS.ByteString (LedgerState era)
-   decodeLedgerState (SerialisedLedgerState (Serialised ls)) =
-     first (const ls) (decodeFull ls)
 
+writeStakeSnapshot :: forall era ledgerera. ()
+  => ShelleyLedgerEra era ~ ledgerera
+  => Era.Era ledgerera
+  => FromCBOR (LedgerState era)
+  => Hash StakePoolKey
+  -> SerialisedLedgerState era
+  -> ExceptT ShelleyQueryCmdError IO ()
+writeStakeSnapshot poolId qState =
+  case decodeLedgerState qState of
+    Left bs -> {- firstExceptT ShelleyQueryCmdHelpersError $ pPrintCBOR bs -} do
+      result <- lift $ runExceptT (pPrintCBOR bs)
+      case result of
+        Left e -> left (ShelleyQueryCmdHelpersError e)
+        Right _ -> return ()
```

Comment:
As far as I can tell, the commented code `firstExceptT ShelleyQueryCmdHelpersError $ pPrintCBOR bs` is equivalent to the expansion below it.

This seems wrong.  We tried to decode the ledger state and failed, and yet there still seems to be a way to return a success.

Reply:
We should have something like:
```
        liftIO $ putTextLn "Version mismatch between node and consensus, so dumping this as generic CBOR."
        firstExceptT ShelleyQueryCmdHelpersError $ pPrintCBOR lbs
        ```

Reply:
Right, but shouldn't it still fail?

Reply:
I don't think it should fail as it's a debugging command and there was never a guarantee that the versions would match. Printing why you're getting a CBOR dump instead of JSON is good enough IMO.

Reply:
Ok cool.  I've restored the original and added an explanatory comment.